### PR TITLE
Curate the  virtual network facts module

### LINKED ISSUE #49885: azure_rm_virtualnetwork deletes subnets when updating custom DNS list
##### SUMMARY
Updating DNS servers list causes removal of Subnets from Virtual Network.
Similar to [39648](https://github.com/ansible/ansible/issues/39648).

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
- azure_rm_virtualnetwork

##### ANSIBLE VERSION
```
ansible 2.7.4
  config file = /Users/michal.golab/.ansible.cfg
  configured module search path = ['/Users/michal.golab/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/michal.golab/.pyenv/versions/3.6.3/envs/ansible_2.7/lib/python3.6/site-packages/ansible
  executable location = /Users/michal.golab/.pyenv/versions/ansible_2.7/bin/ansible
  python version = 3.6.3 (default, Nov 13 2017, 10:36:28) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.38)]
```

##### CONFIGURATION
```
ANSIBLE_NOCOWS(/Users/michal.golab/.ansible.cfg) = True
ANSIBLE_PIPELINING(/Users/michal.golab/.ansible.cfg) = True
ANSIBLE_SSH_ARGS(/Users/michal.golab/.ansible.cfg) = -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=60s
DEFAULT_FORKS(/Users/michal.golab/.ansible.cfg) = 10
DEFAULT_ROLES_PATH(/Users/michal.golab/.ansible.cfg) = ['/Users/michal/ansibleroles']
DEFAULT_SCP_IF_SSH(/Users/michal.golab/.ansible.cfg) = True
DEFAULT_TIMEOUT(/Users/michal.golab/.ansible.cfg) = 20
HOST_KEY_CHECKING(/Users/michal.golab/.ansible.cfg) = False
RETRY_FILES_SAVE_PATH(/Users/michal.golab/.ansible.cfg) = /tmp
```

##### OS / ENVIRONMENT
macOS Mojave
Version: 10.14.2
```
$pip freeze
adal==1.2.0
ansible==2.7.4
ansible-lint==3.4.23
anyconfig==0.9.7
applicationinsights==0.11.7
argcomplete==1.9.4
arrow==0.12.1
asn1crypto==0.24.0
atomicwrites==1.2.1
attrs==18.2.0
azure-cli-core==2.0.35
azure-cli-nspkg==3.0.2
azure-common==1.1.11
azure-graphrbac==0.40.0
azure-keyvault==1.0.0a1
azure-mgmt-batch==4.1.0
azure-mgmt-compute==2.1.0
azure-mgmt-containerinstance==0.4.0
azure-mgmt-containerregistry==2.0.0
azure-mgmt-containerservice==3.0.1
azure-mgmt-dns==1.2.0
azure-mgmt-hdinsight==0.1.0
azure-mgmt-keyvault==0.40.0
azure-mgmt-marketplaceordering==0.1.0
azure-mgmt-monitor==0.5.2
azure-mgmt-network==1.7.1
azure-mgmt-nspkg==2.0.0
azure-mgmt-rdbms==1.2.0
azure-mgmt-resource==1.2.2
azure-mgmt-sql==0.7.1
azure-mgmt-storage==1.5.0
azure-mgmt-trafficmanager==0.50.0
azure-mgmt-web==0.32.0
azure-nspkg==2.0.0
azure-storage==0.35.1
bcrypt==3.1.4
binaryornot==0.4.4
Cerberus==1.2
certifi==2018.10.15
cffi==1.11.5
chardet==3.0.4
click==6.7
click-completion==0.3.1
colorama==0.3.9
cookiecutter==1.6.0
cryptography==2.3.1
dnspython==1.15.0
entrypoints==0.2.3
fasteners==0.14.1
flake8==3.5.0
future==0.17.1
git-url-parse==1.1.0
humanfriendly==4.17
idna==2.7
isodate==0.6.0
Jinja2==2.10
jinja2-time==0.2.0
jmespath==0.9.3
kafka-python==1.4.4
keyring==16.0.0
knack==0.3.3
MarkupSafe==1.0
mccabe==0.6.1
molecule==2.19.0
monotonic==1.5
more-itertools==4.3.0
msrest==0.4.29
msrestazure==0.4.31
oauthlib==2.1.0
packaging==18.0
paramiko==2.4.1
pathspec==0.5.9
pbr==4.1.0
pexpect==4.6.0
pluggy==0.8.0
poyo==0.4.2
psutil==5.4.6
ptyprocess==0.6.0
py==1.7.0
pyasn1==0.4.4
pycodestyle==2.3.1
pycparser==2.19
pyflakes==1.6.0
Pygments==2.2.0
PyJWT==1.6.4
PyNaCl==1.3.0
pyOpenSSL==18.0.0
pyparsing==2.3.0
pytest==3.9.3
python-dateutil==2.7.5
python-gilt==1.2.1
PyYAML==3.13
requests==2.20.0
requests-oauthlib==1.0.0
sh==1.12.14
six==1.11.0
tabulate==0.8.2
testinfra==1.16.0
tree-format==0.1.2
urllib3==1.24
whichcraft==0.5.2
yamllint==1.11.1
```
##### STEPS TO REPRODUCE
1. Create Virtual Network with Subnet and set custom DNS server IP using method of your choice (Web UI, ARM Template, Ansible, etc.)
2. Change custom DNS IP using azure_rm_virtualnetwork module
3. Check if Subnet created earlier still exists

Or just run attached playbook:
```
---
- hosts: localhost
  connection: local
  vars:
    resource_group: resourcegroup
    location: westeurope
    virtual_network: vnet01
  tasks:

  - name: Create Virtual Network
    azure_rm_virtualnetwork:
      resource_group: "{{ resource_group }}"
      name: "{{ virtual_network }}"
      address_prefixes_cidr: 10.1.0.0/16
      dns_servers:
        - 127.0.0.1

  - name: Create Subnet
    azure_rm_subnet:
      name: TEST
      virtual_network_name: "{{ virtual_network }}"
      resource_group: "{{ resource_group }}"
      address_prefix_cidr: 10.1.0.0/24

  - name: Change DNS server
    azure_rm_virtualnetwork:
      resource_group: "{{ resource_group }}"
      name: "{{ virtual_network }}"
      dns_servers:
        - 127.0.0.2
```

The subnet "TEST" will be deleted when playbook finishes (without any error).

If there is already infrastructure within Subnet the result is:
```
TASK [dns : Replace current DNS server in Virtual Network] **************************************************************************************************************************
fatal: [dns01 -> localhost]: FAILED! => {"changed": false, "msg": "Error creating or updating virtual network VNET-DEV-10-WE - Azure Error: InUseSubnetCannotBeDeleted\nMessage: Subnet SUBNET-DEV-10-WE-MGMT is in use by /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/RG-DEV-10-WE/providers/Microsoft.Network/networkInterfaces/DEV-10-WE-MONITOR-NET/ipConfigurations/ipconfig1 and cannot be deleted."}
```
##### EXPECTED RESULTS
IP address of custom DNS will be changed and Subnet will be intact.

##### ACTUAL RESULTS
Subnet is deleted if there is no infrastructure attached to it.
If there is infrastructure attached to Subnet then Ansible session ends with "Azure Error: InUseSubnetCannotBeDeleted".

### PR DESCRIPTION
**Pull request by yuwzho**
##### SUMMARY

fix #49885

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

azure_rm_virtualnetwork_facts

##### ADDITIONAL INFORMATION

```paste below
2.7.0
```

### CODE REVIEW

**Comment by Fred-sun** [comment]
@yungezz @zikalino

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py [hunk 1]
```diff
@@ -82,6 +82,78 @@
         },
         "type": "Microsoft.Network/virtualNetworks"
     }]
+virtualnetworks:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py:85]: with a flag to return curated or raw?

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py:85]: no, the flag was early idea, then it was dropped during one of our Thursday meetings

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py [hunk 2]
```diff
@@ -125,9 +197,14 @@ def exec_module(self, **kwargs):
...
-            self.results['ansible_facts']['azure_virtualnetworks'] = self.get_item()
+            results = self.get_item()
+        elif self.resource_group is not None:
+            results = self.list_resource_group()
         else:
-            self.results['ansible_facts']['azure_virtualnetworks'] = self.list_items()
+            results = self.list_items()
+
+        self.results['ansible_facts']['azure_virtualnetworks'] = self.serialize(results)
+        self.results['ansible_facts']['virtualnetworks'] = self.curated(results)
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py]: ['ansible_facts'] should be removed.
that was also discussed.

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py [hunk 3]
```diff
@@ -169,9 +245,50 @@ def list_items(self):
         results = []
         for item in response:
             if self.has_tags(item.tags, self.tags):
-                results.append(self.serialize_obj(item, AZURE_OBJECT_CLASS))
+                results.append(item)
         return results
 
+    def serialize(self, raws):
+        self.log("Seialize all items")
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py]: Serialize

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py [hunk 4]
```diff
@@ -108,7 +180,7 @@ def __init__(self):
 
         self.results = dict(
             changed=False,
-            ansible_facts=dict(azure_virtualnetworks=[])
+            ansible_facts=dict(azure_virtualnetworks=[], virtualnetworks=[])
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py]: move "virtualnetworks" outside ansible_facts

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py (hunk 1 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py:85]: actually there should be also a comment that raw result is obsolete, and anything non-documented will be removed from version 2.xx

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py (hunk 2 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py]: yes, this what I wanna clarify, please look into this https://github.com/ansible/ansible/issues/49415

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py [hunk 5]
```diff
@@ -82,6 +82,78 @@
...
+                        description:
+                            - Resource Name.
+                        type: str
+                    provisioning_state:
+                        description:
+                            - provision state of the Resource.
+                        type: str
+                        sample: Successed
+                    address_prefix:
+                        description:
```
Comment (mattclay) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py:144]: Missing description here.

Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py:144]: Thanks Matt for figuring out.

**Comment by Fred-sun** [comment]
@yungezz @zikalino

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py [hunk 6]
```diff
@@ -181,7 +181,8 @@ def __init__(self):
 
         self.results = dict(
             changed=False,
-            ansible_facts=dict(azure_virtualnetworks=[], virtualnetworks=[])
+            ansible_facts=dict(azure_virtualnetworks=[]),
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py:184]: docs update?

Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py:184]: it is in the top level of the doc. No need to updated.

Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualnetwork_facts.py:184]: seems i looked at partial change...

**Review by yungezz — APPROVED** [review]

**Comment by yungezz** [comment]
shipit

**Pull request merged**